### PR TITLE
Tahoe invite.3

### PR DESCRIPTION
**Pull request by meejah**
As per https://github.com/gridsync/gridsync/issues/42 this branch updates `tahoe-invite.2` to use a "bare" `wormhole` object and do the client and server "abilities" messages before exchanging the actual configuration.

Note that this actually depends on wormhole "master" as that includes the new `wormhole.wormhole.create()` method among others.

Before merging:

- [x] make `wormhole.tahoe-lafs.org` a CNAME to `wormhole.leastauthority.com`
- [x] depend on a new release of magic-wormhole (with `.get_welcome()` etc)

### CODE REVIEW

**Comment by meejah** [comment]
The only remaining tweak here (besides some squashing and depending on a wormhole release with the new API) would be I think deciding on the APPID and relay-URI to burn into `tahoe` -- we'd talked about putting this at e.g. `wormhole.tahoe-lafs.org`. https://LeastAuthority.com is fine to volunteer to continue to run a wormhole server so this would be a Simple Matter of DNS (right, @exarkun?)

**Comment by crwood** [comment]
For what it's worth, Gridsync currently expects an `appid` of `tahoe-lafs.org/tahoe-lafs/v1` for accepting invites to join a grid (which, to me, seems generic enough). If, for some reason, a different `appid` would be more suitable, however, I would gladly switch to it instead.

Also, +1 to `wormhole.tahoe-lafs.org` for the domain.

**Comment by crwood** [comment]
Would it be more appropriate to make the `magic-wormhole` dependency a `setuptools` "extra" (`pip install`-able via `tahoe-lafs[invite]` or similar) instead of a strict requirement?

**Comment by exarkun** [comment]
> we'd talked about putting this at e.g. wormhole.tahoe-lafs.org. https://LeastAuthority.com is fine to volunteer to continue to run a wormhole server so this would be a Simple Matter of DNS (right, @exarkun?)

I think making wormhole.tahoe-lafs.org a CNAME for wormhole.leastauthority.com (which is already set up and points at a wormhole server) would do it, right?  That should be the work of a couple minutes at most.

**Comment by meejah** [comment]
Hmm. My coverage tool says 100% of the diff is covered...

**Comment by meejah** [comment]
Ah, it does update. but still seems confused? or there's some Heisen-coverage?

**Comment by warner** [comment]
CNAME sounds good to me: tahoe source code will use `tahoe-lafs.org` URLs, but I'm happy with having LAE run the service itself (I figured we'd run it on the same `www.tahoe-lafs.org` host as our Trac server, but I imagine it'll be more reliable if LAE runs it on whatever professional stuff they use).

Maybe we can factor things to make the `tahoe invite` command delegate the work to an internal API, and then the gridsync frontend can `import` the tahoe code and invoke that same internal API. That way we don't need to express *everything* (all the error messages) through a shell command, nor committing to a stable CLI interface, but we avoid the fragility of duplicate code.

**Comment by meejah** [comment]
> Maybe we can factor things to make the tahoe invite command delegate the work to an internal API, and then the gridsync frontend can import the tahoe code and invoke that same internal API. That way we don't need to express everything (all the error messages) through a shell command, nor committing to a stable CLI interface, but we avoid the fragility of duplicate code.

This is a good idea in general, so I'll do this.

However, I don't think it will help GridSync specifically, as it's written in Python3 and tahoe doesn't support that. There's some slim chance we could make "just this one API" able to be imported in py3 which seems fragile.

Also, the API which would help gridsync would be the "client" side (i.e. `tahoe create-node --join <code>`). That said, possibly the Least Authority S4 sign-up side code could benefit from a library version of `tahoe invite`.

The use-cases I expect for `tahoe invite` would be friend-net style grids (i.e. rare, manual inviting), or perhaps "small company" ones. It can also serve as a "lowest common denominator" for other commercial entities to integrate via and also demo code of what the "server" side needs to communicate via wormhole.

Is it worth making a "more machine-readable" version of the output? Perhaps; that could be a future `--json` or similar option?

**Comment by meejah** [comment]
p.s. I think this is worth having in Tahoe before doing any "grid manager" or "accounting" things for at least two reasons:

* it's useful now, on its own
* we can use the wormhole functionality for other tahoe things (immediately: for magic-folder invites and joins) too

File: src/allmydata/_auto_deps.py [hunk 1]
```diff
@@ -91,6 +91,9 @@
     "PyYAML >= 3.11",
 
     "six >= 1.10.0",
+
+    # for 'tahoe invite' and 'tahoe join'
+    "magic-wormhole",  # XXX for now, we need master!
```
Comment (exarkun) [inline src/allmydata/_auto_deps.py]: Are you going to hold off merging this until the next magic-wormhole release?

File: src/allmydata/scripts/create_node.py [hunk 2]
```diff
@@ -344,6 +350,54 @@ def create_node(config):
         os.mkdir(basedir)
     write_tac(basedir, "client")
 
+    # if we're doing magic-wormhole stuff, do it now
+    if config['join'] is not None:
+        print >>out, "Opening wormhole with code '{}'".format(config['join'])
+
+        wh = wormhole.create(
+            appid=u"tahoe-lafs.org/lafs",
+            relay_url=u"ws://wormhole.leastauthority.com:4000/v1",
```
Comment (exarkun) [inline src/allmydata/scripts/create_node.py]: I suppose this is still meant to change - but not until wormhole.tahoe-lafs.org gets configured?  Though I don't see the harm in merging this as-is if setting up that alternate name takes too long.  It would be a simple matter to update it in the future after DNS configuration is done.

File: src/allmydata/scripts/create_node.py [hunk 3]
```diff
@@ -344,6 +350,54 @@ def create_node(config):
...
+            }
+        }
+        wh.send_message(json.dumps(intro))
+
+        server_intro = yield wh.get_message()
+        server_intro = json.loads(server_intro)
+
+        print >>out, "  received server introduction"
+        assert u'abilities' in server_intro
+        assert u'server-v1' in server_intro['abilities']
```
Comment (exarkun) [inline src/allmydata/scripts/create_node.py]: Rather than assert, what do you think about doing a regular `if` and presenting a more presentable failure to the user?  The traceback is likely to be meaningless to 99% of potential users.

Also, Tahoe-LAFS should include some documentation explaining what the `server-v1` ability corresponds to.  This will be pretty boring, I think, but it will set the stage for someone to write something interesting for `server-v2`.

File: src/allmydata/scripts/create_node.py [hunk 4]
```diff
@@ -344,6 +350,54 @@ def create_node(config):
...
+        print >>out, "Overriding the following config:"
+        # presume someone doing "--join" doesn't want to provide
+        # storage; if they do they can change the config after (and/or
+        # allow the 'tahoe invite' to specify storage options?).
+        remote_config['no-storage'] = True
+        remote_config['listen'] = 'none'
+        for k, v in remote_config.items():
+            if k not in ['happy', 'needed', 'total', 'introducer']:
+                print >>out, "   --{}={}".format(k, v)
+            config[k] = v
```
Comment (exarkun) [inline src/allmydata/scripts/create_node.py]: This is a bit confusing.  The config doesn't actually map to `tahoe create` command line arguments, right?  This is more of a metaphor to try to explain things to the user?

I don't see an obvious problem with letting the server any supported configuration option... but part of me still wants to err on the side of a whitelist.  Perhaps because I don't have a deep understanding of every single current Tahoe-LAFS option so I'm not confident it's safe to expose them all to the server.  Or perhaps just out of fear of what new behaviors might arise as folks add new configuration options.

Or, from another angle, if the client lets any option be set, the server won't actually know which options it can set on the client.  If a new config option is added to Tahoe-LAFS but `client-v1`  is still in use, a server will have no way to know whether it's talking to a client that supports that option or not.  It might be harmless to set it for everyone or the server might actually care about doing something different for older clients.  So ... I'd go with a whitelist and bump the client version somehow (either the number or add an item to the corresponding object that tells the server what the client's whitelist is).

Also, the handling of `introducer` points in the direction of another reason for a whitelist.  Perhaps a new sensitive value will be added to the configuration at some point.  Without knowing about each key being set, the client has no way to elide that new sensitive value from the logs.

File: src/allmydata/scripts/tahoe_invite.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+from twisted.python import usage
+from twisted.internet import defer, reactor
+
+from wormhole import wormhole
+
+from allmydata.util import configutil
+
+
+INVITE_SEPARATOR = "+"
```
Comment (exarkun) [inline src/allmydata/scripts/tahoe_invite.py]: This appears to be unused.

File: src/allmydata/scripts/tahoe_invite.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        if len(args) != 1:
+            raise usage.UsageError(
+                "Provide a single argument: the new node's nickname"
+            )
+        self['nick'] = args[0].strip()
+
+
+@defer.inlineCallbacks
+def invite(options):
+    basedir = unicode(options.parent['node-directory'])
```
Comment (exarkun) [inline src/allmydata/scripts/tahoe_invite.py]: `options.parent['node-directory'].decode('ascii')`?  `.decode('utf-8')`?  The former is probably almost exactly the same as `unicode(...)` but it makes the ASCII assumption strongly (Grr PyGtk!) and explicit.

File: src/allmydata/test/cli/test_invite.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+                    u"introducer": u"pb://foo",
+                }),
+            ])
+            w.create = mock.Mock(return_value=fake_wh)
+
+            rc, out, err = yield run_cli(
+                "create-client",
+                "--join", "1-abysmal-ant",
+                node_dir,
+            )
```
Comment (exarkun) [inline src/allmydata/test/cli/test_invite.py]: Hmmm.  I want to suggest making `wormhole` a parameter to all of the APIs that otherwise need it monkey-patched in from the tests.... but `run_cli` seems to get in the way of that idea.  That's too bad. :/

File: src/allmydata/test/cli/test_invite.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+            self.assertEqual(0, rc)
+            with open(join(node_dir, 'tahoe.cfg'), 'r') as f:
+                config = f.read()
+                self.assertTrue(
```
Comment (exarkun) [inline src/allmydata/test/cli/test_invite.py]: From here down could be de-dented.

Comment (exarkun) [inline src/allmydata/test/cli/test_invite.py]: Also, what about `assertIn`?

File: src/allmydata/test/cli/test_invite.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+        yield super(Join, self).setUp()
+        yield self.set_up_grid(oneshare=True)
+
+    @defer.inlineCallbacks
+    def test_create_node_join(self):
+        node_dir = self.mktemp()
+
+        with mock.patch('allmydata.scripts.create_node.wormhole') as w:
+            fake_wh = _create_fake_wormhole([
+                json.dumps({u"abilities": {u"server-v1": {}}}),
```
Comment (exarkun) [inline src/allmydata/test/cli/test_invite.py]: missing, I think, is a test for the behavior when abilities does not contain `u"server-v1"`.

File: src/allmydata/scripts/create_node.py [hunk 10]
```diff
@@ -80,7 +83,7 @@ def validate_where_options(o):
     else:
         # no --location and --port? expect --listen= (maybe the default), and
         # --listen=tcp requires --hostname. But --listen=none is special.
-        if o['listen'] != "none":
+        if o['listen'] != "none" and o.get('join', None) is None:
```
Comment (exarkun) [inline src/allmydata/scripts/create_node.py]: I think this implies that "a dict with some keys, you know what they are, right?" as a data-type is failing `validate_where_options`.  It would be pretty cool to introduce something a little bit better defined here.

Not necessarily part of (or blocking) this PR, since the problem is clearly pre-existing...

File: src/allmydata/scripts/create_node.py [hunk 11]
```diff
@@ -221,7 +225,9 @@ def write_node_config(c, config):
         c.write("\n")
 
     c.write("[node]\n")
-    nickname = argv_to_unicode(config.get("nickname") or "")
+    nickname = config.get("nickname") or u""
+    if not isinstance(nickname, unicode):
```
Comment (exarkun) [inline src/allmydata/scripts/create_node.py]: When is it not?  Is it only for the `u""` default above?  If so, I suggest `b""` as a default and drop the conditional (ie, unconditionally decode).

File: src/allmydata/scripts/create_node.py (hunk 2 above)
Comment (exarkun) [inline src/allmydata/scripts/create_node.py]: Also, these two strings would probably make good global constants somewhere.

File: src/allmydata/scripts/tahoe_invite.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+        "needed": options["needed"] or config.get('client', 'shares.needed'),
+        "total": options["total"] or config.get('client', 'shares.total'),
+        "happy": options["happy"] or config.get('client', 'shares.happy'),
+        "nickname": nick,
+        "introducer": introducer_furl,
+    }
+
+    wh = wormhole.create(
+        appid=u"tahoe-lafs.org/lafs",
+        relay_url=u"ws://wormhole.tahoe-lafs.org:4000/v1",
```
Comment (exarkun) [inline src/allmydata/scripts/tahoe_invite.py]: Oh.  And here is the `wormhole.tahoe-lafs.org` name. :)  That's a good argument in favor of making this a constant somewhere and sharing it with the other application code that needs it.

File: src/allmydata/test/cli/test_invite.py [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+
+            rc, out, err = yield run_cli(
+                "-d", intro_dir,
+                "invite",
+                "--needed", "1",
+                "--happy", "1",
+                "--total", "1",
+                "foo",
+            )
+            self.assertTrue(rc != 0)
```
Comment (exarkun) [inline src/allmydata/test/cli/test_invite.py]: `assertNotEqual`?

File: src/allmydata/test/cli/test_invite.py [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+            self.assertTrue(
```
Comment (exarkun) [inline src/allmydata/test/cli/test_invite.py]: `assertIn`?

File: src/allmydata/scripts/create_node.py (hunk 3 above)
Comment (meejah) [inline src/allmydata/scripts/create_node.py]: yup, good catch

File: src/allmydata/scripts/create_node.py (hunk 4 above)
Comment (meejah) [inline src/allmydata/scripts/create_node.py]: It *does* map to options, but it turns out the only thing *not* in that list is ``--nickname`` so yeah probably just whitelisting it (er, well explicitly pulling out the ones that are ok) is best.

File: src/allmydata/scripts/tahoe_invite.py (hunk 12 above)
Comment (meejah) [inline src/allmydata/scripts/tahoe_invite.py]: yeah :) I now made it a global option. maybe overkill? but might be useful when testing etc...

File: src/allmydata/test/cli/test_invite.py (hunk 7 above)
Comment (meejah) [inline src/allmydata/test/cli/test_invite.py]: Yeah, the whole "runner" and dispatch thing needs to be .. "re-worked".

**Comment by meejah** [comment]
per @warner's comment about API: you *could* currently import the `allmydata.scripts.tahoe_invite` module and run `invite()` -- but you'd need to supply a proper `Options` instance. Is it worth factoring out a method that would basically just be a call-through to turn all the relevant options into kwargs?

Since it makes sense to re-factor/re-work the whole "runner" and dispatch code anyway (in a different PR!) perhaps we could supply a slightly easier way to run any particular command "from python" (which would also help the unit-tests).

**Comment by meejah** [comment]
The protocol for client-v1, server-v1 etc was taken from discussion here: https://github.com/gridsync/gridsync/issues/42

**Comment by meejah** [comment]
Okay, I've squashed and re-based to current master (and changed the URI to `wormhole.tahoe-lafs.org`, optimistically). I believe I've covered all the comments from the review. Thanks, @exarkun !

This still leaves the "what to depend on" part out though!

File: src/allmydata/scripts/tahoe_invite.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+
+from allmydata.util import configutil
+from allmydata.util.encodingutil import argv_to_abspath
+
+
+class InviteOptions(usage.Options):
+    synopsis = "[options] <nickname>"
+    description = "Create a client-only Tahoe-LAFS node (no storage server)."
+
+    optParameters = [
```
Comment (crwood) [inline src/allmydata/scripts/tahoe_invite.py]: For the sake of being consistent with the standard `tahoe create-client`/`tahoe create-node` options, these should probably be changed throughout to `--shares-needed`, `--shares-happy`, and `--shares-total` (instead of just `--needed`, `--happy`, and  `--total`).

**Comment by crwood** [comment]
While testing this, I noticed that `tahoe invite` appears to require that the nodedir be explicitly specified (instead of implicitly selecting and/or falling back onto `~/.tahoe`):
```
(venv2) dev ~ % tahoe invite bob
Traceback (most recent call last):
  File "/home/user/venv2/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 653, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/home/user/venv2/local/lib/python2.7/site-packages/allmydata/scripts/runner.py", line 150, in dispatch
    d = defer.maybeDeferred(f, so)
  File "/home/user/venv2/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 150, in maybeDeferred
    result = f(*args, **kw)
  File "/home/user/venv2/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1532, in unwindGenerator
    return _inlineCallbacks(None, gen, Deferred())
--- <exception caught here> ---
  File "/home/user/venv2/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1386, in _inlineCallbacks
    result = g.send(result)
  File "/home/user/venv2/local/lib/python2.7/site-packages/allmydata/scripts/tahoe_invite.py", line 34, in invite
    basedir = argv_to_abspath(options.parent['node-directory'])
  File "/home/user/venv2/local/lib/python2.7/site-packages/allmydata/util/encodingutil.py", line 104, in argv_to_abspath
    decoded = argv_to_unicode(s)
  File "/home/user/venv2/local/lib/python2.7/site-packages/allmydata/util/encodingutil.py", line 91, in argv_to_unicode
    precondition(isinstance(s, str), s)
  File "/home/user/venv2/local/lib/python2.7/site-packages/allmydata/util/assertutil.py", line 39, in precondition
    raise AssertionError, "".join(msgbuf)
exceptions.AssertionError: precondition: None <type 'NoneType'>
```
Running `tahoe -d ~/.tahoe invite bob` works as expected, however, users arguably shouldn't be required to specify the nodedir if all other commands select `~/.tahoe` by default.

File: src/allmydata/scripts/create_node.py [hunk 16]
```diff
@@ -344,6 +348,74 @@ def create_node(config):
...
     write_tac(basedir, "client")
 
+    # if we're doing magic-wormhole stuff, do it now
+    if config['join'] is not None:
+        print >>out, "Opening wormhole with code '{}'".format(config['join'])
+        relay_url = config.parent['wormhole-server']
+        print >>out, "Connecting to '{}'".format(relay_url)
+
+        wh = wormhole.create(
+            appid=u"tahoe-lafs.org/lafs",
```
Comment (crwood) [inline src/allmydata/scripts/create_node.py]: This should probably be changed to `tahoe-lafs.org/tahoe-lafs/v1` (so as to be compatible with [this](https://github.com/LeastAuthority/leastauthority.com/pull/568/commits/aeed6f45198aee6e0647d11f3b185845a0c8a378#diff-0893f54dd902f12cade21bc3698a4019R341) and others).

On a related note, we might want to consider now whether this should be bumped up to "v2", or even be made more specific (perhaps so as to distinguish the "grid invites" of now from the "magic-folder invites" of the future -- maybe something like `tahoe-lafs.org/tahoe-lafs/invite` vs. `tahoe-lafs.org/tahoe-lafs/magic-folder/invite`?).

**Comment by crwood** [comment]
After further testing, it looks like `create-client --join` step is not actually writing the correct encoding parameters to the newly-created `tahoe.cfg` file (though it does correctly _print_ the parameters)..

To demonstrate, I pointed `wormhole.tahoe-lafs.org` to a working wormhole server in `/etc/hosts`, ran a `tahoe -d ~/.tahoe invite bob` to get an invite code, and then ran `tahoe -d ~/.tahoe-bob create-client --join <invite-code>`. This is the output I saw:

```
Opening wormhole with code '7-hydraulic-flagpole'
Connecting to 'ws://wormhole.tahoe-lafs.org:4000/v1'
Connected to wormhole server
  received server introduction
  received configuration
Encoding: 1 of 1 shares, on at least 1 servers
Overriding the following config:
  listen: none
  introducer: [sensitive data; see tahoe.cfg]
  no-storage: True
  nickname: bob
Node created in '/home/user/.tahoe-bob'
```

Because my `shares.*` values are all set to `1`, the line which reads "Encoding: 1 of 1 shares, on at least 1 servers" is what I would expect to see here, but when I read the newly-created `~/.tahoe-bob/tahoe.cfg` file, I see that those parameters are still set to the defaults:

```
shares.needed = 3
shares.happy = 7
shares.total = 10
```

(The `nickname` and `introducer.furl` values were set correctly, however.)

File: src/allmydata/scripts/create_node.py (hunk 16 above)
Comment (meejah) [inline src/allmydata/scripts/create_node.py]: Did we release anything with the "v1" appid in it?

Also, is the appid the right place to do versioning like this, or in the capabilities exchanged as the first messages? I guess the "pro" of doing it in the appid is no possibilities of "crossing the streams" and accidentally receiving a grid-invite as a magic-folder-invite..? (but still, in such a case the capabilities wouldn't match anyway...)

Comment (meejah) [inline src/allmydata/scripts/create_node.py]: (But yes the main point about the appid being wrong/not-matching above needs to be fixed)

Comment (crwood) [inline src/allmydata/scripts/create_node.py]: >Did we release anything with the "v1" appid in it?

Nope.. I remember @exarkun suggesting earlier, however, that a version-bump might make for good practice (but don't recall us reaching any sort of consensus on moving forward with that idea -- and may even be mistaken in thinking that he was referring to the `appid` specifically now that I think about it..). Personally, I have no qualms either way, but wanted to raise the the question here in the event that anyone felt strongly about what the `appid` should be more generally (since it isn't something we can easily change if/when this PR lands); like you implied, what really matters here is that everything is compatible.

>Also, is the appid the right place to do versioning like this, or in the capabilities exchanged as the first messages?

Hmm.. That's a good question. Reducing the possibility of "crossing the streams" by using separate `appid`s where possible is desirable for both security and usability reasons (since an attacker would need to flood channels on _both_ `appid`s instead of just one -- while fewer nameplates in use means potentially fewer characters to type!) and I was already operating under the assumption that we would be doing this later on to isolate magic-folder-invites from grid-invites (or at least I can't really think of any _good_ reasons right now for keeping the `appid`s the same for both types). On that note, I was under the impression -- and please correct me if I'm wrong -- that the capabilities-exchange was intended more so as a means of handling the situation in which we might want/need to change the data-format of a specific invite-type in a way that is not backwards-compatible with the previous one (i.e., as a means of signaling to users more broadly that they need to update their clients) -- in other words, a means of migrating from grid-invite "version 1" to grid-invite "version 2" independently of the `appid` (and not so much as a means of packing magic-folder-invites in with grid-invites under the same `appid`).

All that being said, I suppose it doesn't _really_ matter at this point, since we're only handling one "type" of invite right now and the capabilities-exchange gives us a good way -- better than changing the `appid`, certainly -- of updating its data-format in the unlikely event that we ever need to (and in retrospect I probably should have framed my original comment to more clearly distinguish between a) the question of isolating future invite-"types" and b) the question of versioning since they can be addressed independently of one another).

Maybe, then, we should just omit the "v1" suffix from the current `appid` altogether (i.e., use `tahoe-lafs.org/tahoe-lafs` _everywhere_) and use the capabilities-exchange to handle any/all "versioning"/update-messaging stuff moving forward? Using the `appid` for versioning seems like a pretty bad idea to me overall (since, in practice, a sender would always have to open at least _two_ wormholes -- one using the "old" `appid`(s) and one using the "new" -- in order to tell clients on the old one to update).. What do you think?

**Comment by meejah** [comment]
Okay I have cleaned up the deficiencies @crwood found, centralized appid handling and changed the option + json naming to `shares-*`. This should now be good-to-go *except* for the CNAME thing that @warner needs to do before merging this for real.

**Comment by meejah** [comment]
re-based this to current master as of today

File: src/allmydata/scripts/tahoe_invite.py [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+                introducer_furl = f.read().strip()
+        except IOError:
+            print >>err, "Can't find introducer fURL in tahoe.cfg nor " + \
+                "{}/private/introducer.furl".format(basedir)
+            raise SystemExit(1)
+
+    nick = options['nick']
+
+    config = {
+        "shares-needed": options["shares-needed"] or config.get('client', 'shares.needed'),
```
Comment (exarkun) [inline src/allmydata/scripts/tahoe_invite.py]: I suppose `0` is not a very interesting or useful value for any of these configuration options .... if it were I could imagine some surprising results from this code.

File: src/allmydata/scripts/tahoe_invite.py [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+        "shares-total": options["shares-total"] or config.get('client', 'shares.total'),
+        "shares-happy": options["shares-happy"] or config.get('client', 'shares.happy'),
+        "nickname": nick,
+        "introducer": introducer_furl,
+    }
+
+    relay_url = options.parent['wormhole-server']
+    print >>out, "Connecting to '{}'...".format(relay_url)
+    wh = wormhole.create(
```
Comment (exarkun) [inline src/allmydata/scripts/tahoe_invite.py]: Similar to above comment, factoring this wormhole interaction out to a separate function seems like it would help the readability here.

File: src/allmydata/scripts/tahoe_invite.py [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+    wh.send_message(json.dumps({
+        u"abilities": {
+            u"server-v1": {},
+        }
+    }))
+
+    client_intro = yield wh.get_message()
+    print >>out, "  received client introduction"
+    client_intro = json.loads(client_intro)
+    if not u'abilities' in client_intro:
```
Comment (exarkun) [inline src/allmydata/scripts/tahoe_invite.py]: `u'abilities' not in client_intro`

File: src/allmydata/test/cli/test_invite.py [hunk 20]
```diff
@@ -0,0 +1,320 @@
...
+
+    @defer.inlineCallbacks
+    def setUp(self):
+        self.basedir = self.mktemp()
+        yield super(Join, self).setUp()
+        yield self.set_up_grid(oneshare=True)
+
+    @defer.inlineCallbacks
+    def test_create_node_join(self):
+        node_dir = self.mktemp()
```
Comment (exarkun) [inline src/allmydata/test/cli/test_invite.py]: Having recently tried to maintain some unit tests with no explanatory documentation regarding the case being covered, I think it would be nice to see some docs for that here.

File: src/allmydata/test/cli/test_invite.py [hunk 21]
```diff
@@ -0,0 +1,320 @@
...
+            rc, out, err = yield run_cli(
+                "-d", intro_dir,
+                "invite",
+                "--shares-needed", "1",
+                "--shares-happy", "1",
+                "--shares-total", "1",
+                "foo",
+            )
+            self.assertNotEqual(rc, 0)
+            self.assertIn(u"Can't find introducer fURL", out + err)
```
Comment (exarkun) [inline src/allmydata/test/cli/test_invite.py]: It seems like the `run_cli` args and following assertions are the only parts which vary between many of these tests.  Perhaps it would be nice to factor out the rest into a helper to reduce future maintenance burden and make the tests themselves easier to read.

File: src/allmydata/test/test_node.py [hunk 22]
```diff
@@ -26,10 +26,14 @@ class TestNode(Node):
     PORTNUMFILE='DEFAULT_PORTNUMFILE_BLANK'
 
 class TestCase(testutil.SignalMixin, unittest.TestCase):
+
+    @defer.inlineCallbacks
     def setUp(self):
         testutil.SignalMixin.setUp(self)
         self.parent = LoggingMultiService()
         self.parent.startService()
+        self._available_port = yield iputil.allocate_tcp_port()
```
Comment (exarkun) [inline src/allmydata/test/test_node.py]: Changes in this module look unrelated, just general reliability improvements for the test suite by avoiding random port collisions?  If so it might be nice to split this off into a separate pr.

File: docs/magic-wormhole-invites.rst [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+friend Bob use it as a client. She runs ``tahoe invite bob`` which
+will print out a short "wormhole code" like ``2-unicorn-quiver``. You
+may also include some options for total, happy and needed shares if
+you like.
+
+Alice then transmits this one-time secret code to Bob. Alice must keep
+her command running until Bob has done his step as it is waiting until
+a secure channel is established before sending the data.
+
+Bob then runs ``tahoe create-client --join <secret code>`` with any
```
Comment (exarkun) [inline docs/magic-wormhole-invites.rst]: I forget if there was a discussion about the verb "join" here.  I think @warner hinted at preferring something like "accept" (but I can't channel his reasoning).  They both make equal sense to me, fwiw.  Either you are "joining" a grid to which you were "invited" or you are "accepting" an "invitation" to a grid...

Comment (meejah) [inline docs/magic-wormhole-invites.rst]: I don't have a good reason for preferring one over the other; happy to paint the bikeshed whichever colour @warner wants :)

File: docs/magic-wormhole-invites.rst [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+
+After receiving the message from the other side and confirming the
+expected protocol, Alice transmits the configuration JSON::
+
+    {
+        "needed": 3,
+        "total": 10,
+        "happy": 7,
+        "nickname": "bob",
+        "introducer": "pb://xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx@example.com:41505"
```
Comment (meejah) [inline docs/magic-wormhole-invites.rst]: thanks, changing

File: docs/running.rst [hunk 25]
```diff
@@ -26,6 +26,19 @@ grid`_ as you only need to create a client node. When you want to create your
 own grid you'll need to create the introducer and several initial storage
 nodes (see the note about small grids below).
 
+
+Being Introduced to a Grid
+--------------------------
+
+A collection of Tahoe servers is called a Grid and usually has 1 Introducer (but sometimes more, and it's possible to run with zero). The Introducer announces which storage servers constitute the Grid and how to contact them. There is a secret "fURL" you need to know to talk to the Introducer.
```
Comment (meejah) [inline docs/running.rst]: hmm, yup. re-wrapped.

File: integration/test_servers_of_happiness.py [hunk 26]
```diff
@@ -24,8 +24,8 @@ def test_upload_immutable(reactor, temp_dir, introducer_furl, flog_gatherer, sto
 
     node_dir = join(temp_dir, 'edna')
 
-    print("waiting 5 seconds unil we're maybe ready")
-    yield task.deferLater(reactor, 5, lambda: None)
+    print("waiting 10 seconds unil we're maybe ready")
+    yield task.deferLater(reactor, 10, lambda: None)
```
Comment (meejah) [inline integration/test_servers_of_happiness.py]: The latter :(

Comment (meejah) [inline integration/test_servers_of_happiness.py]: ...and yes teaching the tests to properly figure out when "the grid is ready" would be the right answer here. Hopefully #430 will help, I'll try (in a new branch).

File: src/allmydata/scripts/tahoe_invite.py [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+    else:
+        basedir = get_default_nodedir()
+    config = configutil.get_config(join(basedir, 'tahoe.cfg'))
+    out = options.stdout
+    err = options.stderr
+
+    try:
+        introducer_furl = config.get('client', 'introducer.furl')
+    except NoSectionError:
+        # we're not a client; maybe this is running *on* the introducer?
```
Comment (meejah) [inline src/allmydata/scripts/tahoe_invite.py]: Yeah, maybe that would be better. There is a helper for that. I don't *really* care what type of node this is, it's more that there isn't just one way to get "the introducer fURL" -- maybe I really want to add a helper for that ...

**Comment by meejah** [comment]
Pulled out the port 1234 stuff: https://github.com/tahoe-lafs/tahoe-lafs/pull/435

**Comment by warner** [comment]
BTW, the CNAME step is done.

**Comment by warner** [comment]
Looks good. I had to restart the travis build for some reason (it completed but then didn't update the ticket). Landing now. Way to go!
